app/src/backend/CalculationService.ts

```diff
@@ -73,70 +81,193 @@ export default class CalculationService {
-          );
-          totalCO2e += co2eq;
-          totalCO2eYears = Math.max(totalCO2eYears, co2eqYears);
-          return { gas: gasName, amount: amount };
-        });
+        gases = CalculationService.handleDirectMeasureFormula(
+          activityValue,
+          gasToCO2Eqs,
+        );
         break;
       case "activity-amount-times-emissions-factor":
-        // TODO add actvityAmount column to ActivityValue
-        // const activityAmount = activityValue.activityAmount || 0;
-        const data = activityValue.activityData;
-        const activityAmount = data ? data["activity_amount"] || 0 : 0;
-        gases = activityValue.gasValues.map((gasValue) => {
-          const emissionsFactor = gasValue.emissionsFactor;
-          if (emissionsFactor == null) {
-            throw new createHttpError.BadRequest(
-              "Missing emissions factor for activity",
-            );
-          }
-          if (emissionsFactor.emissionsPerActivity == null) {
-            throw new createHttpError.BadRequest(
-              `Emissions factor ${emissionsFactor.id} has no emissions per activity`,
-            );
-          }
-          // this rounds/ truncates!
-          const amount = BigInt(
-            activityAmount * emissionsFactor.emissionsPerActivity,
-          );
-          const { co2eq, co2eqYears } = this.calculateCO2eq(
+        gases =
+          CalculationService.handleActivityAmountTimesEmissionsFactorFormula(
+            activityValue,
             gasToCO2Eqs,
-            gasValue.gas!,
-            amount,
           );
-          totalCO2e += co2eq;
-          totalCO2eYears = Math.max(totalCO2eYears, co2eqYears);
-
-          return { gas: gasValue.gas!, amount };
-        });
         break;
+      case "methane-commitment":
+        gases = CalculationService.handleMethaneCommitmentFormula(
+          activityValue,
+          gasToCO2Eqs,
+        );
+        break;
+      case "induced-activity-1": // TODO or VKT-1? What is it in the hierarchy.json file?
+        gases = CalculationService.handleVkt1Formula(
+          activityValue,
+          gasToCO2Eqs,
+        );
       default:
         throw new createHttpError.NotImplemented(
           `Formula ${formula} not yet implemented for input methodology ${inventoryValue.inputMethodology}`,
         );
     }
+
+    for (const gas of gases) {
+      const { co2eq, co2eqYears } = this.calculateCO2eq(
+        gasToCO2Eqs,
+        gas.gas,
+        gas.amount,
+      );
+      totalCO2e += co2eq;
+      totalCO2eYears = Math.max(co2eqYears, totalCO2eYears);
+    }
+
     return {
       totalCO2e,
       totalCO2eYears,
       gases,
     };
   }
+
+  static handleVkt1Formula(
+    activityValue: ActivityValue,
+    gasToCO2Eqs: GasToCO2Eq[],
+  ): Gas[] {
+    const data = activityValue.activityData;
+    if (!data) {
+      throw new createHttpError.BadRequest(
+        "Activity has no data associated, so it can't use the formula",
```

Comment:
how will this message be displayed in the frontend? I'm thinking we need an error code so that it can be translated. Also, we might want a native speaker's POV because to me the sentence sounds a little weird... maybe "No data for this activity, formula can't be applied"

Reply:
Agreed. This is the case for our whole backend, so I'll create a ticket for this so we can apply i18n to backend error messages and display them in the frontend as toasts.

Currently the frontend is handling these errors and then creating custom toast messages from the designs that are translatable.

Reply:
Created ticket for this concerning the whole backend.